### PR TITLE
Use Snyk as data provider

### LINKED ISSUE #717: Snyk used in GitHub projects
The GitHub project uses Snyk for code analysis and sometimes creates PR with the suggestion

Things to do:
- Check if information can be extracted from GitHub PRs, if Snyk is used in the project.
- Check if Snyk configuration exists within the project.

For more information can be found in https://snyk.io/blog/getting-started-snyk-for-secure-python-development/

### PR DESCRIPTION
**Pull request by ManjunathMS35**
Fixes #717

**Changes:**
- Added Snyk as data provider basically to improve the Golang project rating
- Reduced the sub score weightage for the Dependendabot if the GitHub was used for development from 6.0 to 3.0
- With the current change the rating for the project https://github.com/gofiber/fiber has changed from 4.78 to 4.99 

ToDo:
- If the GitHub is used as a development environment is considered in both Dependabot and Snyk due to which both can increase the score value if both tools are used in a single project. Check if this feature can be extracted separately or modify the current implement to consider if the GitHub is used for development form all data providers once and not multiple times.

### CODE REVIEW

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/PackageManagement.java [hunk 1]
```diff
@@ -83,6 +86,7 @@ public class PackageManagement extends CachedSingleFeatureGitHubDataProvider<Pac
         ".vcxproj"::equals, ".fsproj"::equals, "packages.config"::equals);
     register(RUBYGEMS, "Gemfile.lock"::equals, "Gemfile"::equals, ".gemspec"::endsWith);
     register(COMPOSER, "composer.json"::equals, "composer.lock"::equals);
+    register(GOMODULES, "go.mod"::equals);
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/PackageManagement.java:89]: Is this the only form of Go Module possible for Package Management.

Maybe is there a `.lock` file?

File: src/test/java/com/sap/oss/phosphor/fosstars/advice/oss/SnykAdvisorTest.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import com.sap.oss.phosphor.fosstars.model.Rating;
+import com.sap.oss.phosphor.fosstars.model.RatingRepository;
+import com.sap.oss.phosphor.fosstars.model.ValueSet;
+import com.sap.oss.phosphor.fosstars.model.rating.oss.OssSecurityRating;
+import com.sap.oss.phosphor.fosstars.model.subject.oss.GitHubProject;
+import com.sap.oss.phosphor.fosstars.model.value.Languages;
+import com.sap.oss.phosphor.fosstars.model.value.PackageManagers;
+import com.sap.oss.phosphor.fosstars.model.value.ValueHashSet;
+import java.net.MalformedURLException;
+import org.junit.Test;
```
Comment (sourabhsparkala) [inline src/test/java/com/sap/oss/phosphor/fosstars/advice/oss/SnykAdvisorTest.java:33]: Please remove the unused imports

File: src/test/java/com/sap/oss/phosphor/fosstars/advice/oss/SnykAdvisorTest.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    values.update(USES_SNYK.value(false));
+    project.set(rating.calculate(values));
+    assertEquals(1, advisor.adviceFor(project).size());
+    assertEquals("You can create Snyk account and configure your project.",
+        advisor.adviceFor(project).get(0).content().text());
+  }
+
+  @Test
+  public void testAdviceWhenSnykScoreIsNotApplicable() throws MalformedURLException {
+    final DependabotAdvisor advisor = new DependabotAdvisor(WITH_EMPTY_CONTEXT);
```
Comment (sourabhsparkala) [inline src/test/java/com/sap/oss/phosphor/fosstars/advice/oss/SnykAdvisorTest.java:68]:
```suggestion
    final SnykAdvisor advisor = new SnykAdvisor(WITH_EMPTY_CONTEXT);
```

File: src/test/java/com/sap/oss/phosphor/fosstars/advice/oss/SnykAdvisorTest.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+import com.sap.oss.phosphor.fosstars.model.value.PackageManagers;
+import com.sap.oss.phosphor.fosstars.model.value.ValueHashSet;
+import java.net.MalformedURLException;
+import org.junit.Test;
+
+public class SnykAdvisorTest {
+
+  @Test
+  public void testAdviseForSnyk() throws MalformedURLException {
+    SnykAdvisor advisor = new SnykAdvisor(WITH_EMPTY_CONTEXT);
```
Comment (sourabhsparkala) [inline src/test/java/com/sap/oss/phosphor/fosstars/advice/oss/SnykAdvisorTest.java:39]:
```suggestion
    final SnykAdvisor advisor = new SnykAdvisor(WITH_EMPTY_CONTEXT);
```

File: src/test/java/com/sap/oss/phosphor/fosstars/advice/oss/SnykAdvisorTest.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+  @Test
+  public void testAdviceWhenSnykScoreIsNotApplicable() throws MalformedURLException {
+    final DependabotAdvisor advisor = new DependabotAdvisor(WITH_EMPTY_CONTEXT);
+    final GitHubProject project = new GitHubProject("org", "test");
+
+    Rating rating = RatingRepository.INSTANCE.rating(OssSecurityRating.class);
+    ValueSet values = new ValueHashSet();
+    values.update(allUnknown(rating.score().allFeatures()));
+    values.update(USES_SNYK.value(false));
```
Comment (sourabhsparkala) [inline src/test/java/com/sap/oss/phosphor/fosstars/advice/oss/SnykAdvisorTest.java:74]: If this is false, shouldn't the Advisory list contain an advice for Snyk?

And anyways the above test method already tests this scenario, what does this test method actually do?

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/UsesSnyk.java [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+   *
+   * @see <a href="https://github.com/snyk/actions/tree/master/setup">To setup Snyk actions</a>
+   *
+   */
+  private static final String [] SNYK_CONFIGS = {
+      ".github/workflows/snyk.yaml",
+      ".github/workflows/snyk.yml"
+  };
+
+  /** Predicate to confirm if there is a file in open-source project with the .snyk extension. */
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/UsesSnyk.java:60]:
```suggestion
  /** 
    * Predicate to confirm if there is a file in open-source project with the .snyk extension.
    */
```

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/UsesSnyk.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+  /**
+   * A pattern to detect commits by Snyk.
+   *
+   * @see <a
+   *     href="https://docs.snyk.io/integrations/git-repository-scm-integrations/github-integration#commit-signing/">Snyk
+   *     commit signing</a>
+   */
+  private static final String SNYK_PATTERN = "snyk";
+
+  /** Period of time to be checked. */
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/UsesSnyk.java:73]:
```suggestion
  /** 
    * Period of time to be checked.
    */
```

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/UsesSnyk.java [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+  /**
+   * Checks whether a name looks like Snyk.
+   *
+   * @param name The name.
+   * @return True if the name looks like Snyk, false otherwise.
+   */
+  private static boolean isSnyk(String name) {
+    return name != null && name.toLowerCase().contains(SNYK_PATTERN);
+  }
+}
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/UsesSnyk.java:212]: I believe most of the methods given here seems to be from `UsesDependebot`. Could you please provide an `abstract class`?

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/PackageManagement.java (hunk 1 above)
Comment (ManjunathMS35) [reply src/main/java/com/sap/oss/phosphor/fosstars/data/github/PackageManagement.java:89]: In addition to go.mod, the go command maintains a file named go.sum containing the expected [cryptographic hashes](https://go.dev/cmd/go/#hdr-Module_downloading_and_verification) of the content of specific module versions. Do we need to check this as well?

File: src/main/java/com/sap/oss/phosphor/fosstars/model/Score.java [hunk 9]
```diff
@@ -137,4 +137,14 @@ static double adjust(double value) {
...
 
+  /**
+   * Get an Interval with the provided range.
+   *
+   * @param min An interval start value.
+   * @param max An interval end value.
+   * @return Interval with the range provided from min and max param values.
+   */
+  static Interval makeInterval(double min, double max) {
+    return DoubleInterval.init().from(min).to(max).closed().make();
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/model/Score.java:148]: Please add test to address this.

File: src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java [hunk 10]
```diff
@@ -171,6 +171,28 @@ private OssFeatures() {
   public static final BooleanFeature HAS_OPEN_PULL_REQUEST_FROM_DEPENDABOT
       = new BooleanFeature("If a project has open pull requests from Dependabot");
 
+  /**
+   * <p>Shows if a project uses Snyk.</p>
+   * <p><a href="https://snyk.io/">Snyk</a> offers
+   * i) Static Application Security Testing (SAST) amd
+   * i) Static Application Security Testing (SAST) amd
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java:178]:
```suggestion
```

File: src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java [hunk 11]
```diff
@@ -171,6 +171,28 @@ private OssFeatures() {
...
+   * ii) Automatic dependency updates
+   * In particular for automatic dependency updates,
+   * when Snyk finds a vulnerability in dependencies,
+   * it opens a pull request to update the vulnerable dependency to the safe version.</p>
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java:182]: Could you format this a bit better? It looks very congested.

File: src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java [hunk 12]
```diff
@@ -171,6 +171,28 @@ private OssFeatures() {
...
+   */
+  public static final Feature<Boolean> USES_SNYK
+          = new BooleanFeature("If a project uses Snyk");
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java:185]: I guess here indentation seems to be a miss

File: src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java [hunk 13]
```diff
@@ -171,6 +171,28 @@ private OssFeatures() {
...
+
+  /**
+   * Shows if an open source project has open pull requests from Snyk which means that
+   * there are dependencies with known vulnerabilities.
+   *
+   * @see <a href="https://snyk.io/">Snyk</a>
+   */
+  public static final BooleanFeature HAS_OPEN_PULL_REQUEST_FROM_SNYK
+          = new BooleanFeature("If a project has open pull requests from Snyk");
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java:194]: Indentation

File: src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java [hunk 14]
```diff
@@ -171,6 +171,28 @@ private OssFeatures() {
...
+   * it opens a pull request to update the vulnerable dependency to the safe version.</p>
+   */
+  public static final Feature<Boolean> USES_SNYK
+          = new BooleanFeature("If a project uses Snyk");
+
+  /**
+   * Shows if an open source project has open pull requests from Snyk which means that
+   * there are dependencies with known vulnerabilities.
+   *
+   * @see <a href="https://snyk.io/">Snyk</a>
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java:191]:
```suggestion
   * @see <a href="https://snyk.io/">Snyk Introduction</a>
```

File: src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java (hunk 11 above)
Comment (sourabhsparkala) [reply src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java:182]:
```
*<ul>
*   <li></li> 
*</ul>

```

File: src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java (hunk 13 above)
Comment (sourabhsparkala) [reply src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java:194]: This feature seems completely useless, where is it used in SnykScore?

Comment (sourabhsparkala) [reply src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java:194]: I think this is a good feature to increase the case in Snyk Score

File: src/test/java/com/sap/oss/phosphor/fosstars/model/score/oss/SnykDependencyScanScoreTest.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+import static com.sap.oss.phosphor.fosstars.model.feature.oss.OssFeatures.USES_SNYK;
+import static com.sap.oss.phosphor.fosstars.model.other.Utils.setOf;
+import static com.sap.oss.phosphor.fosstars.model.value.Language.GO;
+import static com.sap.oss.phosphor.fosstars.model.value.PackageManager.GOMODULES;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import com.sap.oss.phosphor.fosstars.model.Confidence;
+import com.sap.oss.phosphor.fosstars.model.Score;
+import com.sap.oss.phosphor.fosstars.model.math.DoubleInterval;
```
Comment (sourabhsparkala) [inline src/test/java/com/sap/oss/phosphor/fosstars/model/score/oss/SnykDependencyScanScoreTest.java:17]: Remove this import

File: src/test/resources/com/sap/oss/phosphor/fosstars/model/score/oss/DependabotScoreTestVectors.yml [hunk 16]
```diff
@@ -104,10 +104,10 @@ elements:
         - "JAVA"
   expectedScore:
     type: "DoubleInterval"
-    from: 6.0
+    from: 1.0
     openLeft: false
     negativeInfinity: false
-    to: 9.0
+    to: 3.0
```
Comment (sourabhsparkala) [inline src/test/resources/com/sap/oss/phosphor/fosstars/model/score/oss/DependabotScoreTestVectors.yml:110]: This cannot be changed like this, As the Okay rating comes only between range 5 - 9

File: src/test/resources/com/sap/oss/phosphor/fosstars/model/score/oss/SnykDependencyScanScoreTestVectors.yml [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+        name: "A set of programming languages"
+      languages:
+        elements:
+        - "JAVA"
+  expectedScore:
+    type: "DoubleInterval"
+    from: 1.0
+    openLeft: false
+    negativeInfinity: false
+    to: 3.0
```
Comment (sourabhsparkala) [inline src/test/resources/com/sap/oss/phosphor/fosstars/model/score/oss/SnykDependencyScanScoreTestVectors.yml:110]: This is wrong range for Okay rating

File: src/test/resources/com/sap/oss/phosphor/fosstars/model/score/oss/SnykDependencyScanScoreTestVectors.yml [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+  expectedScore:
+    type: "DoubleInterval"
+    from: 9.0
+    openLeft: false
+    negativeInfinity: false
+    to: 10.0
+    openRight: false
+    positiveInfinity: false
+  expectedLabel: null
+  alias: "very_good"
```
Comment (sourabhsparkala) [inline src/test/resources/com/sap/oss/phosphor/fosstars/model/score/oss/SnykDependencyScanScoreTestVectors.yml:152]: Please add additional test vector for Bad and moderate rating range as well

File: src/test/resources/com/sap/oss/phosphor/fosstars/model/score/oss/VulnerabilityDiscoveryAndSecurityTestingScoreTestVectors.yml [hunk 19]
```diff
@@ -1,5 +1,11 @@
 ---
-defaults: []
+defaults:
+- type: "BooleanValue"
+  feature:
+    type: "BooleanFeature"
+    name: "If a project uses Snyk"
+  flag: false
```
Comment (sourabhsparkala) [inline src/test/resources/com/sap/oss/phosphor/fosstars/model/score/oss/VulnerabilityDiscoveryAndSecurityTestingScoreTestVectors.yml:7]: Maybe this default value is not required, as you are already putting in unknown values

File: src/test/resources/com/sap/oss/phosphor/fosstars/model/rating/oss/OssArtifactSecurityRatingTestVectors.yml [hunk 20]
```diff
@@ -639,6 +648,11 @@ elements:
           type: "BooleanFeature"
           name: "If a project uses Dependabot"
         flag: false
+      - type: "BooleanValue"
+        feature:
+          type: "BooleanFeature"
+          name: "If a project uses Snyk"
+        flag: false
```
Comment (sourabhsparkala) [inline src/test/resources/com/sap/oss/phosphor/fosstars/model/rating/oss/OssArtifactSecurityRatingTestVectors.yml:655]: As a default is already set, maybe this part is not required.

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/UsesSnyk.java (hunk 8 above)
Comment (ManjunathMS35) [reply src/main/java/com/sap/oss/phosphor/fosstars/data/github/UsesSnyk.java:106]: Done

File: src/test/resources/com/sap/oss/phosphor/fosstars/model/score/oss/DependabotScoreTestVectors.yml [hunk 21]
```diff
@@ -104,10 +104,10 @@ elements:
         - "JAVA"
   expectedScore:
     type: "DoubleInterval"
-    from: 6.0
+    from: 5
```
Comment (sourabhsparkala) [inline src/test/resources/com/sap/oss/phosphor/fosstars/model/score/oss/DependabotScoreTestVectors.yml:107]:
```suggestion
    from: 5.0
```

File: src/test/resources/com/sap/oss/phosphor/fosstars/model/score/oss/DependabotScoreTestVectors.yml [hunk 22]
```diff
@@ -104,10 +104,10 @@ elements:
...
     openLeft: false
     negativeInfinity: false
-    to: 9.0
+    to: 8
```
Comment (sourabhsparkala) [inline src/test/resources/com/sap/oss/phosphor/fosstars/model/score/oss/DependabotScoreTestVectors.yml:110]:
```suggestion
    to: 8.0
```

File: src/test/resources/com/sap/oss/phosphor/fosstars/model/score/oss/SnykDependencyScanScoreTestVectors.yml [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+    - type: "LanguagesValue"
+      feature:
+        type: "LanguagesFeature"
+        name: "A set of programming languages"
+      languages:
+        elements:
+        - "JAVA"
+  expectedScore:
+    type: "DoubleInterval"
+    from: 5
```
Comment (sourabhsparkala) [inline src/test/resources/com/sap/oss/phosphor/fosstars/model/score/oss/SnykDependencyScanScoreTestVectors.yml:107]:
```suggestion
    from: 5.0
```

File: src/test/resources/com/sap/oss/phosphor/fosstars/model/score/oss/SnykDependencyScanScoreTestVectors.yml [hunk 24]
```diff
@@ -0,0 +1,152 @@
...
+    openLeft: false
+    negativeInfinity: false
+    to: 8
```
Comment (sourabhsparkala) [inline src/test/resources/com/sap/oss/phosphor/fosstars/model/score/oss/SnykDependencyScanScoreTestVectors.yml:110]:
```suggestion
    to: 8.0
```

**Review by sourabhsparkala — APPROVED** [review]
👍

**Pull request merged**